### PR TITLE
Load PEFT weights from local directory

### LINKED ISSUE #1259: PEFT support does not work with local directories
### System Info

text-generation-inference version: main branch

The rest of the system info is probably not relevant.

### Information

- [X] Docker
- [X] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Reproduction

Steps to reproduce:

1. Create a directory that contains PEFT weights, for example: `mypeft` which contains two files: `adapter_config.json` and `adapter_model.bin`
2. Try to download the base model using the following command: `text-generation-server download-weights mypeft`

### Expected behavior

It should be able to download the base model weights and merge the PEFT weights into the base model.

However, as mentioned in [this comment](https://github.com/huggingface/text-generation-inference/pull/762#issuecomment-1728042436):

> it seems the checks for PEFT model is only in the block checking if the model is not a local model

There is a workaround as documented in [this comment](https://github.com/huggingface/text-generation-inference/pull/762#issuecomment-1729689087)

### PR DESCRIPTION
**Pull request by tleyden**
# What does this PR do?

Enables PEFT weights to be loaded from a local directory, as opposed to a hf hub repository.  It is a continuation of the work in PR https://github.com/huggingface/text-generation-inference/pull/762

Fixes #1259 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?  **Yes but I don't know how to run the tests for this repo, and it doesn't look like this code is covered anyway**
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.  **Yes, @Narsil asked for a PR in [this comment](https://github.com/huggingface/text-generation-inference/pull/762#issuecomment-1728089505)**
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).  **I didn't see any documentation added to the [original PR](https://github.com/huggingface/text-generation-inference/pull/762), and am not sure where this belongs.  Let me know and I can add some**
- [x] Did you write any new necessary tests?  **I didn't see any existing test coverage for this python module**

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@Narsil

### CODE REVIEW

File: server/text_generation_server/cli.py [hunk 1]
```diff
@@ -150,6 +150,16 @@ def download_weights(
...
 
+    # Try to load as a local PEFT model
+    try:
+        utils.download_and_unload_peft(
+            model_id, revision, trust_remote_code=trust_remote_code
+        )
+        utils.weight_files(model_id, revision, extension)
+        return
+    except (utils.LocalEntryNotFoundError, utils.EntryNotFoundError):
+        pass
```
Comment (Narsil) [inline server/text_generation_server/cli.py:161]:
```suggestion
    else:
        # Try to load as a local PEFT model
        try:
            utils.download_and_unload_peft(
                model_id, revision, trust_remote_code=trust_remote_code
            )
            utils.weight_files(model_id, revision, extension)
            return
        except (utils.LocalEntryNotFoundError, utils.EntryNotFoundError):
            pass
```

**Review by Narsil — APPROVED** [review]
LGTM

**Comment by echotong1** [comment]
It seems like a bug which lead to launch failed, error info looks like:
  File "/usr/local/bin/text-generation-server", line 8, in <module>
    sys.exit(app())

  File "/text-generation-inference-1.3.2/server/text_generation_server/cli.py", line 204, in download_weights
    utils.download_and_unload_peft(

  File "/text-generation-inference-1.3.2/server/text_generation_server/utils/peft.py", line 24, in download_and_unload_peft
    model = AutoPeftModelForSeq2SeqLM.from_pretrained(

  File "/usr/local/lib/python3.9/site-packages/peft/auto.py", line 69, in from_pretrained
    peft_config = PeftConfig.from_pretrained(pretrained_model_name_or_path, **kwargs)

  File "/usr/local/lib/python3.9/site-packages/peft/utils/config.py", line 121, in from_pretrained
    raise ValueError(f"Can't find '{CONFIG_NAME}' at '{pretrained_model_name_or_path}'")

**Comment by tleyden** [comment]
@echotong1 this pr already merged.  Can you open a separate issue with steps to reproduce and expected vs actual behavior?

**Pull request merged**